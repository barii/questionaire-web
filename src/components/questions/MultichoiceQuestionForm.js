import React from 'react';
import TextFieldGroup from '../common/TextFieldGroup';
import ListFieldGroup from '../common/ListFieldGroup';
import { Form, Button, Card, CardHeader, CardBody } from 'reactstrap';


export default function MultichoiceQuestionForm({question, errors, onChange, onSubmit, onAdd}) {
  return (
    <div>
    <Card>
      <CardHeader>Header</CardHeader>
      <CardBody>
        <Form onSubmit={onSubmit} > 
          <TextFieldGroup
            label="Text"
            name="text"
            value={question.text}
            onChange={onChange}
            error={errors.text}
          />
          <TextFieldGroup
            label="Description"
            name="description"
            value={question.description}
            onChange={onChange}
            error={errors.description}
          />
          <ListFieldGroup
            label="Choices"
            name="choices"
            options={question.choices}
            onChange={onChange}
            error={errors.choices}
          />
          <div>
            <Button id="btnAdd" color="primary" onClick={onAdd}>Add</Button>
          </div>
          <div>
            <Button type="submit" color="success">Save</Button>
          </div>
        </Form>
      </CardBody>
    </Card>
  </div>
  );
}
