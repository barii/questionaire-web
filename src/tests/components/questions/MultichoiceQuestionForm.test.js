import React from 'react';
import { shallow } from 'enzyme';
import MultichoiceQuestionForm from '../../../components/questions/MultichoiceQuestionForm';

const question = {
  type: "MultichoiceQuestion",
  id: '1',
  text: "haha",
  description: "desc",
  choices: [
    {
      id: 1,
      value: "Choice 1"
    }
  ]
}

const onSubmit = jest.fn()
const onChange = jest.fn()
const onAdd = jest.fn()

const wrapper = shallow(
  <MultichoiceQuestionForm 
    question={question} 
    errors
    onChange={onChange} 
    onSubmit={onSubmit}
    onAdd={onAdd} 
  />
);

it('should render MultichoiceQuestionForm correctly (snapshot)', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should call `onSubmit` on button click', () => {
  wrapper.find('Form').simulate('submit');
  expect(onSubmit).toHaveBeenCalled();
});

it('should call `onChange` on value update', () => {
  wrapper.find('TextFieldGroup[name="text"]').simulate('change');
  expect(onChange).toHaveBeenCalled();
});

it('should call `onChange` on description update', () => {
  wrapper.find('TextFieldGroup[name="description"]').simulate('change');
  expect(onChange).toHaveBeenCalled();
});

it('should call `onChange` on choice update', () => {
  wrapper.find('TextFieldGroup[name="description"]').simulate('change');
  expect(onChange).toHaveBeenCalled();
});

it('should call `onAdd` when `Add` button is clicked', () => {
  wrapper.find('#btnAdd').simulate('click');
  expect(onAdd).toHaveBeenCalled();
});
