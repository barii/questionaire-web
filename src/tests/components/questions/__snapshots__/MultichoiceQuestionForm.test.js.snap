// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MultichoiceQuestionForm correctly (snapshot) 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MultichoiceQuestionForm
    errors={true}
    onAdd={[Function]}
    onChange={[Function]}
    onSubmit={[Function]}
    question={
        Object {
            "choices": Array [
              Object {
                "id": 1,
                "value": "Choice 1",
              },
            ],
            "description": "desc",
            "id": "1",
            "text": "haha",
            "type": "MultichoiceQuestion",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Card
        tag="div"
>
        <CardHeader
                tag="div"
        >
                Header
        </CardHeader>
        <CardBody
                tag="div"
        >
                <Form
                        onSubmit={[Function]}
                        tag="form"
                >
                        <TextFieldGroup
                                error={undefined}
                                label="Text"
                                name="text"
                                onChange={[Function]}
                                type="text"
                                value="haha"
                        />
                        <TextFieldGroup
                                error={undefined}
                                label="Description"
                                name="description"
                                onChange={[Function]}
                                type="text"
                                value="desc"
                        />
                        <ListFieldGroup
                                error={undefined}
                                label="Choices"
                                name="choices"
                                onChange={[Function]}
                                options={
                                        Array [
                                                Object {
                                                  "id": 1,
                                                  "value": "Choice 1",
                                                },
                                              ]
                                }
                        />
                        <div>
                                <Button
                                        color="primary"
                                        id="btnAdd"
                                        onClick={[Function]}
                                        tag="button"
                                >
                                        Add
                                </Button>
                        </div>
                        <div>
                                <Button
                                        color="success"
                                        tag="button"
                                        type="submit"
                                >
                                        Save
                                </Button>
                        </div>
                </Form>
        </CardBody>
</Card>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CardHeader
            tag="div"
>
            Header
</CardHeader>,
          <CardBody
            tag="div"
>
            <Form
                        onSubmit={[Function]}
                        tag="form"
            >
                        <TextFieldGroup
                                    error={undefined}
                                    label="Text"
                                    name="text"
                                    onChange={[Function]}
                                    type="text"
                                    value="haha"
                        />
                        <TextFieldGroup
                                    error={undefined}
                                    label="Description"
                                    name="description"
                                    onChange={[Function]}
                                    type="text"
                                    value="desc"
                        />
                        <ListFieldGroup
                                    error={undefined}
                                    label="Choices"
                                    name="choices"
                                    onChange={[Function]}
                                    options={
                                                Array [
                                                            Object {
                                                              "id": 1,
                                                              "value": "Choice 1",
                                                            },
                                                          ]
                                    }
                        />
                        <div>
                                    <Button
                                                color="primary"
                                                id="btnAdd"
                                                onClick={[Function]}
                                                tag="button"
                                    >
                                                Add
                                    </Button>
                        </div>
                        <div>
                                    <Button
                                                color="success"
                                                tag="button"
                                                type="submit"
                                    >
                                                Save
                                    </Button>
                        </div>
            </Form>
</CardBody>,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Header",
            "tag": "div",
          },
          "ref": null,
          "rendered": "Header",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Form
              onSubmit={[Function]}
              tag="form"
>
              <TextFieldGroup
                            error={undefined}
                            label="Text"
                            name="text"
                            onChange={[Function]}
                            type="text"
                            value="haha"
              />
              <TextFieldGroup
                            error={undefined}
                            label="Description"
                            name="description"
                            onChange={[Function]}
                            type="text"
                            value="desc"
              />
              <ListFieldGroup
                            error={undefined}
                            label="Choices"
                            name="choices"
                            onChange={[Function]}
                            options={
                                          Array [
                                                        Object {
                                                          "id": 1,
                                                          "value": "Choice 1",
                                                        },
                                                      ]
                            }
              />
              <div>
                            <Button
                                          color="primary"
                                          id="btnAdd"
                                          onClick={[Function]}
                                          tag="button"
                            >
                                          Add
                            </Button>
              </div>
              <div>
                            <Button
                                          color="success"
                                          tag="button"
                                          type="submit"
                            >
                                          Save
                            </Button>
              </div>
</Form>,
            "tag": "div",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <TextFieldGroup
                  error={undefined}
                  label="Text"
                  name="text"
                  onChange={[Function]}
                  type="text"
                  value="haha"
/>,
                <TextFieldGroup
                  error={undefined}
                  label="Description"
                  name="description"
                  onChange={[Function]}
                  type="text"
                  value="desc"
/>,
                <ListFieldGroup
                  error={undefined}
                  label="Choices"
                  name="choices"
                  onChange={[Function]}
                  options={
                                    Array [
                                                      Object {
                                                        "id": 1,
                                                        "value": "Choice 1",
                                                      },
                                                    ]
                  }
/>,
                <div>
                  <Button
                                    color="primary"
                                    id="btnAdd"
                                    onClick={[Function]}
                                    tag="button"
                  >
                                    Add
                  </Button>
</div>,
                <div>
                  <Button
                                    color="success"
                                    tag="button"
                                    type="submit"
                  >
                                    Save
                  </Button>
</div>,
              ],
              "onSubmit": [Function],
              "tag": "form",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "error": undefined,
                  "label": "Text",
                  "name": "text",
                  "onChange": [Function],
                  "type": "text",
                  "value": "haha",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "error": undefined,
                  "label": "Description",
                  "name": "description",
                  "onChange": [Function],
                  "type": "text",
                  "value": "desc",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "error": undefined,
                  "label": "Choices",
                  "name": "choices",
                  "onChange": [Function],
                  "options": Array [
                    Object {
                      "id": 1,
                      "value": "Choice 1",
                    },
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Button
                    color="primary"
                    id="btnAdd"
                    onClick={[Function]}
                    tag="button"
>
                    Add
</Button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Add",
                    "color": "primary",
                    "id": "btnAdd",
                    "onClick": [Function],
                    "tag": "button",
                  },
                  "ref": null,
                  "rendered": "Add",
                  "type": [Function],
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Button
                    color="success"
                    tag="button"
                    type="submit"
>
                    Save
</Button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Save",
                    "color": "success",
                    "tag": "button",
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": "Save",
                  "type": [Function],
                },
                "type": "div",
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Card
          tag="div"
>
          <CardHeader
                    tag="div"
          >
                    Header
          </CardHeader>
          <CardBody
                    tag="div"
          >
                    <Form
                              onSubmit={[Function]}
                              tag="form"
                    >
                              <TextFieldGroup
                                        error={undefined}
                                        label="Text"
                                        name="text"
                                        onChange={[Function]}
                                        type="text"
                                        value="haha"
                              />
                              <TextFieldGroup
                                        error={undefined}
                                        label="Description"
                                        name="description"
                                        onChange={[Function]}
                                        type="text"
                                        value="desc"
                              />
                              <ListFieldGroup
                                        error={undefined}
                                        label="Choices"
                                        name="choices"
                                        onChange={[Function]}
                                        options={
                                                  Array [
                                                            Object {
                                                              "id": 1,
                                                              "value": "Choice 1",
                                                            },
                                                          ]
                                        }
                              />
                              <div>
                                        <Button
                                                  color="primary"
                                                  id="btnAdd"
                                                  onClick={[Function]}
                                                  tag="button"
                                        >
                                                  Add
                                        </Button>
                              </div>
                              <div>
                                        <Button
                                                  color="success"
                                                  tag="button"
                                                  type="submit"
                                        >
                                                  Save
                                        </Button>
                              </div>
                    </Form>
          </CardBody>
</Card>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <CardHeader
              tag="div"
>
              Header
</CardHeader>,
            <CardBody
              tag="div"
>
              <Form
                            onSubmit={[Function]}
                            tag="form"
              >
                            <TextFieldGroup
                                          error={undefined}
                                          label="Text"
                                          name="text"
                                          onChange={[Function]}
                                          type="text"
                                          value="haha"
                            />
                            <TextFieldGroup
                                          error={undefined}
                                          label="Description"
                                          name="description"
                                          onChange={[Function]}
                                          type="text"
                                          value="desc"
                            />
                            <ListFieldGroup
                                          error={undefined}
                                          label="Choices"
                                          name="choices"
                                          onChange={[Function]}
                                          options={
                                                        Array [
                                                                      Object {
                                                                        "id": 1,
                                                                        "value": "Choice 1",
                                                                      },
                                                                    ]
                                          }
                            />
                            <div>
                                          <Button
                                                        color="primary"
                                                        id="btnAdd"
                                                        onClick={[Function]}
                                                        tag="button"
                                          >
                                                        Add
                                          </Button>
                            </div>
                            <div>
                                          <Button
                                                        color="success"
                                                        tag="button"
                                                        type="submit"
                                          >
                                                        Save
                                          </Button>
                            </div>
              </Form>
</CardBody>,
          ],
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Header",
              "tag": "div",
            },
            "ref": null,
            "rendered": "Header",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Form
                onSubmit={[Function]}
                tag="form"
>
                <TextFieldGroup
                                error={undefined}
                                label="Text"
                                name="text"
                                onChange={[Function]}
                                type="text"
                                value="haha"
                />
                <TextFieldGroup
                                error={undefined}
                                label="Description"
                                name="description"
                                onChange={[Function]}
                                type="text"
                                value="desc"
                />
                <ListFieldGroup
                                error={undefined}
                                label="Choices"
                                name="choices"
                                onChange={[Function]}
                                options={
                                                Array [
                                                                Object {
                                                                  "id": 1,
                                                                  "value": "Choice 1",
                                                                },
                                                              ]
                                }
                />
                <div>
                                <Button
                                                color="primary"
                                                id="btnAdd"
                                                onClick={[Function]}
                                                tag="button"
                                >
                                                Add
                                </Button>
                </div>
                <div>
                                <Button
                                                color="success"
                                                tag="button"
                                                type="submit"
                                >
                                                Save
                                </Button>
                </div>
</Form>,
              "tag": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <TextFieldGroup
                    error={undefined}
                    label="Text"
                    name="text"
                    onChange={[Function]}
                    type="text"
                    value="haha"
/>,
                  <TextFieldGroup
                    error={undefined}
                    label="Description"
                    name="description"
                    onChange={[Function]}
                    type="text"
                    value="desc"
/>,
                  <ListFieldGroup
                    error={undefined}
                    label="Choices"
                    name="choices"
                    onChange={[Function]}
                    options={
                                        Array [
                                                            Object {
                                                              "id": 1,
                                                              "value": "Choice 1",
                                                            },
                                                          ]
                    }
/>,
                  <div>
                    <Button
                                        color="primary"
                                        id="btnAdd"
                                        onClick={[Function]}
                                        tag="button"
                    >
                                        Add
                    </Button>
</div>,
                  <div>
                    <Button
                                        color="success"
                                        tag="button"
                                        type="submit"
                    >
                                        Save
                    </Button>
</div>,
                ],
                "onSubmit": [Function],
                "tag": "form",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "error": undefined,
                    "label": "Text",
                    "name": "text",
                    "onChange": [Function],
                    "type": "text",
                    "value": "haha",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "error": undefined,
                    "label": "Description",
                    "name": "description",
                    "onChange": [Function],
                    "type": "text",
                    "value": "desc",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "error": undefined,
                    "label": "Choices",
                    "name": "choices",
                    "onChange": [Function],
                    "options": Array [
                      Object {
                        "id": 1,
                        "value": "Choice 1",
                      },
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Button
                      color="primary"
                      id="btnAdd"
                      onClick={[Function]}
                      tag="button"
>
                      Add
</Button>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Add",
                      "color": "primary",
                      "id": "btnAdd",
                      "onClick": [Function],
                      "tag": "button",
                    },
                    "ref": null,
                    "rendered": "Add",
                    "type": [Function],
                  },
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Button
                      color="success"
                      tag="button"
                      type="submit"
>
                      Save
</Button>,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Save",
                      "color": "success",
                      "tag": "button",
                      "type": "submit",
                    },
                    "ref": null,
                    "rendered": "Save",
                    "type": [Function],
                  },
                  "type": "div",
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
